crates/executor/state_handler/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    fn kv_set(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+        let mut vec = Vec::new(); //TODO allocations occur here
+        vec.extend_from_slice(value);
+        self.snapshot_state.set(constructed_key, &vec);
+        Ok(())
+    }
+
+    fn kv_delete(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+        self.snapshot_state.delete(&constructed_key);
+        Ok(())
+    }
+
+    fn accumulator_get(
+        &self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+    ) -> Result<u128, StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, true);
+
+        match self.snapshot_state.get(&constructed_key, Global) {
+            Some(value) => {
+                let mut data = [0u8; 16];
+                data.copy_from_slice(&value);
+                Ok(u128::from_be_bytes(data))
+            }
+            None => Ok(0),
+        }
+    }
+
+    fn accumulator_add(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: u128,
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, true);
+
+        let bz = self.kv_get(account_id, scope, key, Global)?;
+
+        let old_value: u128 = match bz {
+            Some(value) => {
+                let mut data = [0u8; 16];
+                data.copy_from_slice(&value);
+                let ov: u128 = u128::from_be_bytes(data);
+                ov
+            }
+            None => 0,
+        };
+
+        let new_value = old_value.add(value);
+
+        let mut vec = Vec::new();
+        vec.extend_from_slice(&new_value.to_le_bytes());
+        self.snapshot_state.set(constructed_key, &vec);
+        Ok(())
+    }
+
+    fn accumulator_safe_sub(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: u128,
+    ) -> Result<bool, StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, true);
+        let bz = self.kv_get(account_id, scope, key, Global)?;
+
+        let old_value: u128 = match bz {
+            Some(value) => {
+                let mut data = [0u8; 16];
+                data.copy_from_slice(&value);
+                let ov: u128 = u128::from_be_bytes(data);
+                ov
+            }
+            None => 0,
+        };
+
+        #[allow(clippy::manual_saturating_arithmetic)]
+        let new_value = old_value.checked_sub(value).unwrap_or(0);
```

Comment:
Actually if `checked_sub` fails then we want to return a custom error code, probably `ErrorCode::HandlerCode(0)`. What clippy is saying is that you're implementing saturating sub manually. But that's not what we want, we want to fail if subtraction will take us below zero.

Reply:
added error, can you check when you have a moment